### DISCUSSION TITLE
safe_reraise() raison d'etre please?

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#7918] [Usage Questions]
I've noticed the use of safe_reraise() in SQLA, and read the discussion at #2703 .

Is the co-routine library behavior being mitigated by safe_reraise() a concern in "modern" versions of co-routine libraries?

In other words, is safe_reraise() mostly necessary to mitigate against bad behaviors in older version of co-routine libraries?

Thanks in advance for the education!

**Comment by zzzeek** [comment]
that's a super old issue but re-reading the description, the thing it describes seems very likely to still be the case, but it may be worth trying to reproduce.    Interestingly, in our current mainbranch I recall recently adding a comment where I'm asking the same question, but reading that issue I think I'm going to update that comment.    

Basically, if you're in eventlet/gevent when context switches occur on IO, if you are inside of an `except:` block, and you then call `.rollback` which does IO, you will get a context switch.    In that context switch, anything can happen, most certainly that sys.exc_info() can change.  so the context manager is storing that traceback, weathering out the potential IO/context switches, then re-raising with that context in place.   I don't know that `greenlet` (which is what all these libraries use) is going to have anything in place to change that.
